File: ios/FamWake/Services/Scheduler.swift
import Foundation

// MARK: - Scheduler
// 1:1 Port von shared/src/commonMain/.../algorithm/Scheduler.kt

struct Scheduler {

    func calculateIdealSchedule(
        members: [FamilyMember],
        breakfastDurationMinutes: Int = 30,
        globalBufferMinutes: Int = 0
    ) -> FamilySchedule {

        let activeMembers = members.filter { !$0.isPaused }.prefix(6).map { $0 }

        if activeMembers.isEmpty {
            return FamilySchedule(memberSchedules: [], breakfastTime: nil, isValid: true, scheduleMessage: .noActiveSchedule)
        }

        // 1. Versuche die exakte Reihung ohne Zeit-Verschiebung, mit vollem Puffer
        let initialResult = evaluatePermutation(activeMembers, breakfastDurationMinutes: breakfastDurationMinutes, shiftMinutes: 0, globalBufferMinutes: globalBufferMinutes, includeInvalid: true)
        if initialResult.isValid { return initialResult }

        // 2. Puffer schrittweise reduzieren (5er-Schritte)
        if globalBufferMinutes > 0 {
            var reducedBuffer = globalBufferMinutes - 5
            while reducedBuffer >= 0 {
                let bufferResult = evaluatePermutation(activeMembers, breakfastDurationMinutes: breakfastDurationMinutes, shiftMinutes: 0, globalBufferMinutes: reducedBuffer)
                if bufferResult.isValid {
                    var finalRes = bufferResult
                    finalRes.scheduleMessage = .bufferReduced(globalBufferMinutes, reducedBuffer)
                    return finalRes
                }
                reducedBuffer -= 5
            }
        }

        // 3. Fallback: Erlaube moderate Zeit-Verschiebungen (5-15 Min) bei festgehaltener Reihung, ohne Puffer
        for shift in stride(from: 5, through: 15, by: 5) {
            let flexibleSchedule = evaluatePermutation(activeMembers, breakfastDurationMinutes: breakfastDurationMinutes, shiftMinutes: shift, globalBufferMinutes: 0)
            if flexibleSchedule.isValid {
                var finalRes = flexibleSchedule
                finalRes.scheduleMessage = .timeAdjusted(shift)
                return finalRes
            }
        }

        // 4. Fallback: Frühstück leicht verkürzen und mit Verschiebungen erneut probieren
        if breakfastDurationMinutes >= 15 {
            for reduceBreakfast in stride(from: 5, through: 10, by: 5) {
                let reducedDuration = breakfastDurationMinutes - reduceBreakfast
                let reductionSchedule = evaluatePermutation(activeMembers, breakfastDurationMinutes: reducedDuration, shiftMinutes: 0, globalBufferMinutes: 0)

                if reductionSchedule.isValid {
                    var finalRes = reductionSchedule
                    finalRes.scheduleMessage = .breakfastReduced(reduceBreakfast)
                    return finalRes
                }

                // Kombiniere Frühstücksverkürzung mit Zeit-Verschiebung
                for shift in stride(from: 5, through: 15, by: 5) {
                    let flexibleReductionSchedule = evaluatePermutation(activeMembers, breakfastDurationMinutes: reducedDuration, shiftMinutes: shift, globalBufferMinutes: 0)
                    if flexibleReductionSchedule.isValid {
                        var finalRes = flexibleReductionSchedule
                        finalRes.scheduleMessage = .breakfastAndTimeAdjusted(reduceBreakfast, shift)
                        return finalRes
                    }
                }
            }
        }

        // 5. Endgültiger Fehlschlag: Best-Effort Plan zurückgeben (vom ersten Versuch)
        var failedResult = initialResult
        failedResult.isValid = false
        failedResult.scheduleMessage = extractConflictMessage(initialResult)
        return failedResult
    }

    // MARK: - Private

    private func extractConflictMessage(_ schedule: FamilySchedule) -> ScheduleMessage {
        for s in schedule.memberSchedules {
            if s.wakeUpTime < s.member.earliestWakeUp {
                return .memberConflict(s.member.name)
            }
        }
        return .memberConflict("")
    }

    private func evaluatePermutation(
        _ orderedMembers: [FamilyMember],
        breakfastDurationMinutes: Int,
        shiftMinutes: Int,
        globalBufferMinutes: Int,
        includeInvalid: Bool = false
    ) -> FamilySchedule {

        let breakfastEaters = orderedMembers.filter { $0.wantsBreakfast && !$0.isSimpleMode }
        var breakfastTime: DateComponents? = nil

        if !breakfastEaters.isEmpty {
            var minLeaveForBreakfastEaters = DateComponents(hour: 23, minute: 59)
            for m in breakfastEaters {
                let naturalBathEnd = m.latestWakeUp.adding(minutes: m.bathroomDurationMinutes)
                let leave = m.leaveHomeTime ?? naturalBathEnd
                if leave < minLeaveForBreakfastEaters {
                    minLeaveForBreakfastEaters = leave
                }
            }
            
            // Safety-Guard: Clamp auf 04:00
            let startTime = minLeaveForBreakfastEaters.totalMinutes < 4 * 60 ? DateComponents(hour: 4, minute: 0) : minLeaveForBreakfastEaters
            breakfastTime = startTime.subtracting(minutes: breakfastDurationMinutes)
            
            // Wraparound Check (03:30)
            if let bt = breakfastTime, startTime < bt {
                breakfastTime = DateComponents(hour: 3, minute: 30)
            }
        }

        var schedules: [MemberSchedule] = []
        var currentLatestBathroomEnd = DateComponents(hour: 23, minute: 59)
        var isValid = true

        for index in stride(from: orderedMembers.count - 1, through: 0, by: -1) {
            let member = orderedMembers[index]
            if member.isSimpleMode {
                let wakeUpTime = member.latestWakeUp
                schedules.append(MemberSchedule(
                    member: member,
                    wakeUpTime: wakeUpTime,
                    bathroomStart: wakeUpTime,
                    bathroomEnd: wakeUpTime,
                    bufferAfter: 0
                ))
                continue
            }

            let allowedLatest = member.latestWakeUp.adding(minutes: shiftMinutes)
            let allowedEarliest = member.earliestWakeUp.subtracting(minutes: shiftMinutes)

            // Fixierte Weckzeit (z.B. durch Snooze): earliestWakeUp == latestWakeUp
            let isFixed = member.earliestWakeUp == member.latestWakeUp

            var maxBathroomEnd = allowedLatest.adding(minutes: member.bathroomDurationMinutes)

            if currentLatestBathroomEnd < maxBathroomEnd {
                maxBathroomEnd = currentLatestBathroomEnd
            }

            if member.wantsBreakfast, let bt = breakfastTime, maxBathroomEnd.totalMinutes >= bt.totalMinutes {
                maxBathroomEnd = bt
            }

            if let leave = member.leaveHomeTime, leave < maxBathroomEnd {
                maxBathroomEnd = leave
            }

            // Bei fixierter Weckzeit: exakt diese Zeit verwenden
            let wakeUpTime: DateComponents
            let bathroomEnd: DateComponents
            if isFixed {
                wakeUpTime = member.latestWakeUp
                bathroomEnd = wakeUpTime.adding(minutes: member.bathroomDurationMinutes)
            } else {
                wakeUpTime = maxBathroomEnd.subtracting(minutes: member.bathroomDurationMinutes)
                bathroomEnd = maxBathroomEnd
            }

            if !isFixed && wakeUpTime < allowedEarliest {
                if !includeInvalid {
                    return FamilySchedule(memberSchedules: [], breakfastTime: nil, isValid: false, scheduleMessage: .memberConflict(""))
                }
                isValid = false
            }

            // Der Puffer nach diesem Mitglied bestimmt den Abstand zum Nachfolgenden.
            // Wenn es kein nachfolgendes Mitglied gibt (letztes Element), ist der Puffer 0.
            var effectiveBuffer = 0
            if index < orderedMembers.count - 1 {
                // dayProfiles enthält nach resolveEffectiveMember() nur 1 Entry (den aktuellen Wochentag),
                // daher ist .first hier korrekt und liefert immer das richtige Tagesprofil.
                let activeProfile = member.dayProfiles?.sorted(by: { $0.key < $1.key }).first?.value
                effectiveBuffer = activeProfile?.bufferMinutes ?? globalBufferMinutes
            }

            // Für den nächsten Schritt (das vorhergehende Mitglied) ziehen wir den Puffer
            // des vorhergehenden Mitglieds von unserer Wakeup-Zeit ab.
            var prevBuffer = 0
            if index > 0 {
                let prevMember = orderedMembers[index - 1]
                if !prevMember.isSimpleMode {
                    // dayProfiles enthält nach resolveEffectiveMember() nur 1 Entry (den aktuellen Wochentag),
                    // daher ist .first hier korrekt und liefert immer das richtige Tagesprofil.
                    let activeProfile = prevMember.dayProfiles?.sorted(by: { $0.key < $1.key }).first?.value
                    prevBuffer = activeProfile?.bufferMinutes ?? globalBufferMinutes
                }
            }

            schedules.append(MemberSchedule(
                member: member,
                wakeUpTime: wakeUpTime,
                bathroomStart: wakeUpTime,
                bathroomEnd: bathroomEnd,
                bufferAfter: effectiveBuffer
            ))
            
            currentLatestBathroomEnd = wakeUpTime.subtracting(minutes: prevBuffer)
        }

        // Vorwärts-Korrektur-Pass: Wenn ein fixierter Member (Snooze) seine Badzeit
        // in den Slot des Nachfolgers schiebt, werden nachfolgende Members verschoben.
        var forwardSchedules = Array(schedules.reversed())
        for i in 0..<(forwardSchedules.count - 1) {
            let current = forwardSchedules[i]
            let next = forwardSchedules[i + 1]
            if next.member.isSimpleMode { continue }

            let buffer = current.bufferAfter
            let requiredNextStart = current.bathroomEnd.adding(minutes: buffer)

            if next.wakeUpTime < requiredNextStart {
                let shiftedWakeUp = requiredNextStart
                let shiftedBathroomEnd = shiftedWakeUp.adding(minutes: next.member.bathroomDurationMinutes)
                forwardSchedules[i + 1] = MemberSchedule(
                    member: next.member,
                    wakeUpTime: shiftedWakeUp,
                    bathroomStart: shiftedWakeUp,
                    bathroomEnd: shiftedBathroomEnd,
                    bufferAfter: next.bufferAfter
                )
            }
        }

        // Prüfe ob ein Snooze (fixierter Member) aktiv ist – dann toleriere Verschiebungen
        let hasSnoozeActive = orderedMembers.contains { $0.earliestWakeUp == $0.latestWakeUp }

        // Post-Validation Frühstück (bei Snooze tolerieren, da temporär)
        if let bt = breakfastTime, isValid, !hasSnoozeActive {
            for s in forwardSchedules {
                if s.member.wantsBreakfast && bt < s.bathroomEnd {
                    isValid = false
                    if !includeInvalid {
                        return FamilySchedule(memberSchedules: [], breakfastTime: nil, isValid: false, scheduleMessage: .memberConflict(""))
                    }
                }
            }
        }

        // Bei Snooze-bedingter Verschiebung: Plan als gültig markieren, da die
        // Verschiebung temporär ist und kein User-Eingriff erforderlich ist.
        if hasSnoozeActive && !isValid {
            isValid = true
        }

        return FamilySchedule(
            memberSchedules: forwardSchedules,
            breakfastTime: breakfastTime,
            isValid: isValid,
            scheduleMessage: isValid ? .optimal : .memberConflict("")
        )
    }
}

// MARK: - DateComponents Arithmetic Helpers (Scheduler)
private extension DateComponents {
    func adding(minutes: Int) -> DateComponents {
        let total = self.totalMinutes + minutes
        return DateComponents(hour: total / 60, minute: total % 60)
    }

    func subtracting(minutes: Int) -> DateComponents {
        var total = self.totalMinutes - minutes
        // M10: Mitternachts-Wraparound – bei negativer Differenz auf den Vortag wrappen
        // statt auf 0 zu clampen, z.B. 01:00 - 90min = 23:30 statt 00:00
        if total < 0 { total += 24 * 60 }
        return DateComponents(hour: total / 60, minute: total % 60)
    }
}

File: shared/src/commonMain/kotlin/de/familienwecker/famwake/algorithm/Scheduler.kt
package de.familienwecker.famwake.algorithm

import de.familienwecker.famwake.model.FamilyMember
import de.familienwecker.famwake.model.FamilySchedule
import de.familienwecker.famwake.model.ScheduleMessage
import de.familienwecker.famwake.model.ScheduleResult
import de.familienwecker.famwake.util.minusMinutes
import de.familienwecker.famwake.util.plusMinutes
import de.familienwecker.famwake.util.isBefore
import de.familienwecker.famwake.util.isAfter
import kotlinx.datetime.LocalTime

class Scheduler {

    fun calculateIdealSchedule(
        members: List<FamilyMember>,
        breakfastDurationMinutes: Long = 30,
        globalBufferMinutes: Long = 0
    ): FamilySchedule {
        // Limit active members (max 6 per UI constraint)
        val activeMembers = members.filter { !it.isPaused }.take(6)

        if (activeMembers.isEmpty()) {
            return FamilySchedule(emptyList(), null, true, ScheduleMessage.NoActiveMembers)
        }

        // 1. Versuche die exakte Reihung ohne Zeit-Verschiebung, mit vollem Puffer
        val initialResult = evaluatePermutation(activeMembers, breakfastDurationMinutes, 0, globalBufferMinutes, includeInvalid = true)
        if (initialResult.isValid) return initialResult

        // 2. Puffer schrittweise reduzieren (5er-Schritte)
        if (globalBufferMinutes > 0) {
            var reducedBuffer = globalBufferMinutes - 5
            while (reducedBuffer >= 0) {
                val bufferResult = evaluatePermutation(activeMembers, breakfastDurationMinutes, 0, reducedBuffer)
                if (bufferResult.isValid) {
                    return bufferResult.copy(scheduleMessage = ScheduleMessage.BufferReduced(globalBufferMinutes, reducedBuffer))
                }
                reducedBuffer -= 5
            }
        }

        // 3. Fallback: Erlaube moderate Zeit-Verschiebungen (5-15 Min) bei festgehaltener Reihung, ohne Puffer
        for (shiftMinutes in 5..15 step 5) {
            val flexibleSchedule = evaluatePermutation(activeMembers, breakfastDurationMinutes, shiftMinutes, 0)
            if (flexibleSchedule.isValid) {
                return flexibleSchedule.copy(scheduleMessage = ScheduleMessage.TimeAdjusted(shiftMinutes))
            }
        }

        // 4. Fallback: Frühstück leicht verkürzen und mit Verschiebungen erneut probieren
        if (breakfastDurationMinutes >= 15) {
            for (reduceBreakfast in 5..10 step 5) {
                val reducedDuration = breakfastDurationMinutes - reduceBreakfast
                val reductionSchedule = evaluatePermutation(activeMembers, reducedDuration, 0, 0)

                if (reductionSchedule.isValid) {
                    return reductionSchedule.copy(scheduleMessage = ScheduleMessage.BreakfastReduced(reduceBreakfast))
                }

                // Kombiniere Frühstücksverkürzung mit Zeit-Verschiebung
                for (shiftMinutes in 5..15 step 5) {
                    val flexibleReductionSchedule = evaluatePermutation(activeMembers, reducedDuration, shiftMinutes, 0)
                    if (flexibleReductionSchedule.isValid) {
                        return flexibleReductionSchedule.copy(scheduleMessage = ScheduleMessage.BreakfastAndTimeAdjusted(reduceBreakfast, shiftMinutes))
                    }
                }
            }
        }

        // 5. Endgültiger Fehlschlag: Best-Effort Plan zurückgeben (vom ersten Versuch)
        return initialResult.copy(
            isValid = false,
            scheduleMessage = extractConflictMessage(initialResult)
        )
    }

    private fun extractConflictMessage(schedule: FamilySchedule): ScheduleMessage {
        // Findet das erste Mitglied, das den Check gerissen hat (wakeUpTime < earliestWakeUp)
        for (s in schedule.memberSchedules) {
            if (s.wakeUpTime.isBefore(s.member.earliestWakeUp)) {
                return ScheduleMessage.MemberConflict(s.member.name)
            }
        }
        return ScheduleMessage.NoValidScheduleFound
    }

    private fun evaluatePermutation(
        orderedMembers: List<FamilyMember>,
        breakfastDurationMinutes: Long,
        shiftToleranceMinutes: Int = 0,
        globalBufferMinutes: Long = 0,
        includeInvalid: Boolean = false
    ): FamilySchedule {
        val breakfastEaters = orderedMembers.filter { it.wantsBreakfast && !it.isSimpleMode }
        var breakfastTime: LocalTime? = null

        if (breakfastEaters.isNotEmpty()) {
            var minLeaveForBreakfastEaters = LocalTime(23, 59)
            for (m in breakfastEaters) {
                val naturalBathEnd = m.latestWakeUp.plusMinutes(m.bathroomDurationMinutes)
                val leave = m.leaveHomeTime ?: naturalBathEnd
                if (leave.isBefore(minLeaveForBreakfastEaters)) {
                    minLeaveForBreakfastEaters = leave
                }
            }
            // Safety-Guard: Wenn leaveHomeTime oder naturalBathEnd vor 04:00 liegt
            // (z.B. durch Mitternachts-Wrap-Around der plusMinutes-Arithmetik),
            // auf 04:00 clampen. Verhindert unsinnige Frühstückszeiten um 2 Uhr nachts.
            val startTime = if (minLeaveForBreakfastEaters.isBefore(LocalTime(4, 0)))
                LocalTime(4, 0) else minLeaveForBreakfastEaters

            breakfastTime = startTime.minusMinutes(breakfastDurationMinutes)
            // Clamp auch breakfastTime: startTime ist auf 04:00 geclampt, aber
            // breakfastTime = 04:00 - Dauer kann trotzdem in die Nacht rutschen.
            if (breakfastTime?.isAfter(startTime) == true) {
                // plusMinutes-Wraparound aufgetreten → auf 03:30 als hartes Minimum setzen
                breakfastTime = LocalTime(3, 30)
            }
        }

        val schedules = mutableListOf<ScheduleResult>()
        var currentLatestBathroomEndTime = LocalTime(23, 59)
        var isValid = true

        for (index in orderedMembers.indices.reversed()) {
            val member = orderedMembers[index]
            if (member.isSimpleMode) {
                val wakeUpTime = member.latestWakeUp
                schedules.add(
                    ScheduleResult(
                        member = member,
                        wakeUpTime = wakeUpTime,
                        bathroomStartTime = wakeUpTime,
                        bathroomEndTime = wakeUpTime,
                        bufferAfter = 0L
                    )
                )
                continue
            }

            val allowedLatestWakeUp = member.latestWakeUp.plusMinutes(shiftToleranceMinutes.toLong())
            val allowedEarliestWakeUp = member.earliestWakeUp.minusMinutes(shiftToleranceMinutes.toLong())

            // Fixierte Weckzeit (z.B. durch Snooze): earliestWakeUp == latestWakeUp
            // → Weckzeit direkt verwenden, Constraint an Vorgänger weitergeben
            val isFixed = member.earliestWakeUp == member.latestWakeUp

            var maxAllowedBathroomEnd = allowedLatestWakeUp.plusMinutes(member.bathroomDurationMinutes)

            if (currentLatestBathroomEndTime.isBefore(maxAllowedBathroomEnd)) {
                maxAllowedBathroomEnd = currentLatestBathroomEndTime
            }

            if (member.wantsBreakfast && breakfastTime != null && !breakfastTime.isAfter(maxAllowedBathroomEnd)) {
                maxAllowedBathroomEnd = breakfastTime
            }

            val leaveTime = member.leaveHomeTime
            if (leaveTime != null && leaveTime.isBefore(maxAllowedBathroomEnd)) {
                maxAllowedBathroomEnd = leaveTime
            }

            // Bei fixierter Weckzeit: exakt diese Zeit verwenden, Badende daraus berechnen
            val wakeUpTime: LocalTime
            val bathroomEnd: LocalTime
            if (isFixed) {
                wakeUpTime = member.latestWakeUp
                bathroomEnd = wakeUpTime.plusMinutes(member.bathroomDurationMinutes)
            } else {
                wakeUpTime = maxAllowedBathroomEnd.minusMinutes(member.bathroomDurationMinutes)
                bathroomEnd = maxAllowedBathroomEnd
            }

            if (!isFixed && wakeUpTime.isBefore(allowedEarliestWakeUp)) {
                if (!includeInvalid) {
                    return FamilySchedule(emptyList(), null, false, ScheduleMessage.NoValidScheduleFound)
                }
                isValid = false
            }

            // Der Puffer nach diesem Mitglied bestimmt den Abstand zum Nachfolgenden.
            // Wenn es kein nachfolgendes Mitglied gibt (letztes Element), ist der Puffer 0.
            val effectiveBuffer = if (index < orderedMembers.lastIndex) {
                member.dayProfiles?.values?.firstOrNull()?.bufferMinutes ?: globalBufferMinutes
            } else {
                0L
            }

            // Für den nächsten Schritt (das vorhergehende Mitglied) ziehen wir den Puffer
            // des vorhergehenden Mitglieds von unserer Wakeup-Zeit ab.
            val prevBuffer = if (index > 0) {
                val prevMember = orderedMembers[index - 1]
                if (prevMember.isSimpleMode) {
                    0L
                } else {
                    prevMember.dayProfiles?.values?.firstOrNull()?.bufferMinutes ?: globalBufferMinutes
                }
            } else {
                0L
            }

            schedules.add(
                ScheduleResult(
                    member = member,
                    wakeUpTime = wakeUpTime,
                    bathroomStartTime = wakeUpTime,
                    bathroomEndTime = bathroomEnd,
                    bufferAfter = effectiveBuffer
                )
            )
            currentLatestBathroomEndTime = wakeUpTime.minusMinutes(prevBuffer)
        }

        // Vorwärts-Korrektur-Pass: Wenn ein fixierter Member (Snooze) seine Badzeit
        // in den Slot des Nachfolgers schiebt, müssen nachfolgende Members nach hinten
        // verschoben werden. schedules ist hier noch reversed (letzter Member zuerst).
        val forwardSchedules = schedules.reversed().toMutableList()
        for (i in 0 until forwardSchedules.size - 1) {
            val current = forwardSchedules[i]
            val next = forwardSchedules[i + 1]
            if (next.member.isSimpleMode) continue

            val buffer = current.bufferAfter
            val requiredNextStart = current.bathroomEndTime.plusMinutes(buffer)

            // Wenn der nachfolgende Member vor dem Ende des aktuellen (+ Puffer) startet → verschieben
            if (next.wakeUpTime.isBefore(requiredNextStart)) {
                val shiftedWakeUp = requiredNextStart
                val shiftedBathroomEnd = shiftedWakeUp.plusMinutes(next.member.bathroomDurationMinutes)
                forwardSchedules[i + 1] = next.copy(
                    wakeUpTime = shiftedWakeUp,
                    bathroomStartTime = shiftedWakeUp,
                    bathroomEndTime = shiftedBathroomEnd
                )
            }
        }



        // Prüfe ob ein Snooze (fixierter Member) aktiv ist – dann toleriere Verschiebungen
        val hasSnoozeActive = orderedMembers.any { it.earliestWakeUp == it.latestWakeUp }

        // Post-Validation (auf korrigierte Schedules) – bei Snooze tolerieren, da temporär
        if (breakfastTime != null && isValid && !hasSnoozeActive) {
            for (s in forwardSchedules) {
                if (s.member.wantsBreakfast && s.bathroomEndTime.isAfter(breakfastTime)) {
                    isValid = false
                    if (!includeInvalid) return FamilySchedule(emptyList(), null, false, ScheduleMessage.NoValidScheduleFound)
                }
            }
        }

        // Bei Snooze-bedingter Verschiebung: Plan als gültig markieren, da die
        // Verschiebung temporär ist und kein User-Eingriff erforderlich ist.
        if (hasSnoozeActive && !isValid) {
            isValid = true
        }

        return FamilySchedule(
            memberSchedules = forwardSchedules,
            breakfastTime = breakfastTime,
            isValid = isValid,
            scheduleMessage = if (isValid) ScheduleMessage.OptimalPlan else ScheduleMessage.NoValidScheduleFound
        )
    }
}
